scio-core/src/main/scala/com/spotify/scio/coders/CoderMaterializer.scala

```diff
@@ -73,29 +75,37 @@ object CoderMaterializer {
       case Fallback(_) =>
         new KryoAtomicCoder[T](o.kryo)
       case Transform(c, f) =>
-        val uc = f(beamImpl(o, c))
-        beamImpl(o, uc)
+        val uc = f(beamImpl(o, c, refs))
+        beamImpl(o, uc, refs)
       case Record(typeName, coders, construct, destruct) =>
         RecordCoder(
           typeName,
-          coders.map { case (n, c) => n -> beamImpl(o, c) },
+          coders.map { case (n, c) => n -> beamImpl(o, c, refs) },
           construct,
           destruct
         )
       case Disjunction(typeName, idCoder, id, coders) =>
         DisjunctionCoder(
           typeName,
-          beamImpl(o, idCoder),
+          beamImpl(o, idCoder, refs),
           id,
-          coders.map { case (k, u) => k -> beamImpl(o, u) }
+          coders.map { case (k, u) => k -> beamImpl(o, u, refs) }
         )
       case KVCoder(koder, voder) =>
         // propagate topLevel to k & v coders
-        val kbc = beamImpl(o, koder, topLevel)
-        val vbc = beamImpl(o, voder, topLevel)
+        val kbc = beamImpl(o, koder, refs, topLevel)
+        val vbc = beamImpl(o, voder, refs, topLevel)
         KvCoder.of(kbc, vbc)
-      case Ref(t, c) =>
-        LazyCoder[T](t, o)(c)
+      case r @ Ref(t, c) =>
+        refs.get(r) match {
+          case Some(rc) =>
+            LazyCoder(t, rc.bcoder.asInstanceOf[BCoder[T]])
+          case None =>
+            val rc = RefCoder[T]()
+            refs += r -> rc
+            rc.bcoder = beamImpl(o, c, refs)
+            rc
```

Comment:
inspired from file history. See #2605 
@jto was there an issue with this ?

Reply:
The PR adds a couple of new tests so I suppose that was the issue :)